README.md
# Jira to Json

A Python library to convert Jira issues to Json.

[![GitHub branch checks state](https://img.shields.io/github/checks-status/bharel/jira_to_json/master)](https://github.com/bharel/jira_to_json/actions)
[![PyPI](https://img.shields.io/pypi/v/jira2json)](https://pypi.org/project/jira2json/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/jira2json)](https://pypi.org/project/jira2json/)
[![codecov](https://codecov.io/gh/bharel/jira_to_json/branch/master/graph/badge.svg?token=37IZCOYI9U)](https://codecov.io/gh/bharel/jira_to_json)

The library is based on the [Jira REST API](https://docs.atlassian.com/jira/REST/latest/),
and uses the [Requests](http://docs.python-requests.org/en/latest/) library.

It contacts the Jira server, downloads the issues according to the specified
JQL query, and converts them to Json. An API is provided to access the issues
and their fields and for further processing.

Supports Jira Datacenter.

## Installation

Install the library with pip:

`pip install "jira2json[dotenv]"`

The `dotenv` extra installs the [python-dotenv](https://pypi.org/project/python-dotenv/)
library, which is used to load the Jira server's url and the authentication token
from a `.env` file.

## Usage

The library can be used as a command line tool or as a Python library.

### Command line tool

The command line tool is called `jira2json` and is installed with the library.

Run `jira2json --help` for usage information.

### Python library

The library exports the following functions:

* `iterate_jira_issues`: iterates over the issues returned by a JQL query.
* `prase_issues`: applies parsers on the issue's data.
* `save_jsons_to_file`: saves the issues to a file.

Typical usage:

```python

from jira2json import iterate_jira_issues, prase_issues, save_jsons_to_file

# Iterate over the issues returned by the JQL query
issues = iterate_jira_issues(
    base_url='https://jira.atlassian.com',
    jql='project=JRA',
    token='mytoken',
)

# Parse the jsons and convert them to a more usable format
issues = prase_issues(issues)

# Save the issues to a file
with open('issues.json', 'w') as f:
    save_jsons_to_file(issues, f)
```

## Development

Install the `dev` extra:

`pip install -e .[dev,dotenv]`


changelog.md
# CHANGELOG.md

## 0.1.0 (2023-05-20)

First release!


tests.py
from collections import deque
from dataclasses import dataclass
import io
from threading import Thread
from types import SimpleNamespace
from unittest import TestCase
from unittest.mock import patch
import urllib.parse
import os
from jira2json.__main__ import _main, _parse_args, DEFAULT_FILENAME
from http.server import BaseHTTPRequestHandler, HTTPServer
import weakref

import requests

from jira2json import parse_issues, iterate_jira_issues, save_jsons_to_file
import jira2json


class TestParsers(TestCase):
    def test_worklog_parser(self):
        """Test the worklog parser."""

        result = parse_issues([{
            "Log Work": "comment;started;author;timeSpentSeconds"}])

        self.assertEqual(next(result)["Log Work"], [{
            "comment": "comment", "started": "started", "author": "author",
            "timeSpentSeconds": "timeSpentSeconds"}])

        result = parse_issues([{
            "Log Work": ["comment;started;author;timeSpentSeconds",
                         "comment2;started2;author2;timeSpentSeconds2"]}])

        self.assertEqual(next(result)["Log Work"], [{
            "comment": "comment", "started": "started", "author": "author",
            "timeSpentSeconds": "timeSpentSeconds"},
            {"comment": "comment2", "started": "started2", "author": "author2",
                "timeSpentSeconds": "timeSpentSeconds2"}])

    def test_comment_parser(self):
        """Test the comment parser."""

        # Comments sometimes have trailing semicolons
        result = parse_issues([{
            "Comment": "datetime;author;comment;;;"}])

        self.assertEqual(next(result)["Comment"], [{
            "datetime": "datetime", "author": "author",
            "comment": "comment;;;"}])

        result = parse_issues([{
            "Comment": ["datetime;author;comment",
                        "datetime2;author2;comment2"]}])

        self.assertEqual(next(result)["Comment"], [{
            "datetime": "datetime", "author": "author", "comment": "comment"},
            {"datetime": "datetime2", "author": "author2",
             "comment": "comment2"}])

    def test_override_parsers(self):
        """Test overriding the default parsers."""

        def parser(value):
            return value

        result = parse_issues([{"Log Work": "value"}],
                              parsers={"Log Work": parser})

        self.assertEqual(next(result)["Log Work"], "value")

        with self.assertRaises(StopIteration):
            next(result)


class TestSaveJson(TestCase):
    def test_save_json(self):
        """Test saving JSON to a file."""

        data = [{"key": "value"}]

        fileobj = io.StringIO()
        save_jsons_to_file(data, fileobj)
        fileobj.seek(0)

        self.assertEqual(fileobj.read(), '{"key": "value"}\n')


class FakeHTTPServer:  # pragma: no cover
    @dataclass
    class Request:
        method: str
        path: str
        headers: dict[str, str]
        params: dict[str, list[str]]

    def __init__(self) -> None:
        self.responses: deque[tuple[int, str]] = deque()
        self.requests: deque[self.Request] = deque()  # type: ignore
        get_func = weakref.WeakMethod(self._do_GET)

        class Handler(BaseHTTPRequestHandler):
            # Override GET method
            def do_GET(self):
                func = get_func()
                if func is None:
                    raise RuntimeError("Fake server has been shut down")
                return func(self)

            # Disable logging
            def log_message(self, format, *args):
                pass

        self._server = HTTPServer(("localhost", 0), Handler)
        self.url = f"http://localhost:{self._server.server_port}"
        self._thread = Thread(target=self._server.serve_forever, daemon=True)

    def __enter__(self) -> "FakeHTTPServer":
        self.start()
        return self

    def start(self) -> None:
        self._thread.start()

    def stop(self) -> None:
        self._server.shutdown()
        self._thread.join(5)
        if self._thread.is_alive():
            raise RuntimeError("Fake server did not shut down")

    def __exit__(self, exc_type, exc_value, traceback) -> None:
        self.stop()

    def _do_GET(self, handler):
        url = urllib.parse.urlparse(handler.path)
        query = urllib.parse.parse_qs(url.query)
        headers = dict(handler.headers)
        self.requests.append(
            self.Request(handler.command, url.path, headers, query))
        code, data = self.responses.popleft()
        handler.send_response(code)
        handler.send_header("Content-Type", "text/csv")
        handler.send_header("Content-Length", str(len(data)))
        handler.end_headers()
        handler.wfile.write(data.encode("utf-8"))


class TestIterateJiraIssues(TestCase):
    fake_server: FakeHTTPServer

    @classmethod
    def setUpClass(cls) -> None:
        cls.fake_server = FakeHTTPServer()
        cls.fake_server.start()
        cls.addClassCleanup(cls.fake_server.stop)

    def setUp(self) -> None:
        self.fake_server.requests.clear()
        self.fake_server.responses.clear()
        return super().setUp()

    def test_iterate_jira_issues(self):
        """Test iterating over JIRA issues."""
        self.fake_server.responses.append(
            (200, 'Key,Summary\n"hello","world"'))
        iterator = iterate_jira_issues(self.fake_server.url, "hello",
                                       token="mytoken")
        issue = next(iterator)
        self.assertEqual(issue["Key"], "hello")
        self.assertEqual(issue["Summary"], "world")
        request = self.fake_server.requests.popleft()
        self.assertEqual(request.method, "GET")
        self.assertEqual(request.path, jira2json.CSV_API_PATH)
        self.assertEqual(request.headers["Authorization"], "Bearer mytoken")

        self.assertEqual(request.params["jqlQuery"], ["hello"])
        self.assertEqual(request.params["tempMax"],
                         [str(jira2json.BATCH_SIZE)])
        self.assertEqual(request.params["pager/start"], ["0"])

    def test_changing_headers(self):
        """Headers change between the CSVs"""
        self.fake_server.responses.append(
            (200, 'Key,Summary\n' + '"hello","world"\n'*jira2json.BATCH_SIZE))
        self.fake_server.responses.append((200, 'Header1,Header2\n' +
                                           '"more","data"\n'*10))

        iterator = iterate_jira_issues(self.fake_server.url, "hello",
                                       token="mytoken")

        [next(iterator) for _ in range(jira2json.BATCH_SIZE)]

        with self.assertLogs("jira2json", level="WARNING") as logs:
            next(iterator)

        self.assertEqual(len(logs.records), 1)

        # One issue was inside the log message
        self.assertEqual(sum(1 for _ in iterator), 9)

    def test_longer_csv(self):
        """CSV is longer than the batch size"""
        self.fake_server.responses.append(
            (200, 'Key,Summary\n' +
             '"hello","world"\n'*(jira2json.BATCH_SIZE+1)))

        iterator = iterate_jira_issues(self.fake_server.url, "hello",
                                       token="mytoken")

        with self.assertRaises(AssertionError):
            sum(1 for _ in iterator)

    def test_manual_session(self):
        """A session is given instead of a token"""
        self.fake_server.responses.append((200, ""))  # Empty CSV

        session = requests.Session()
        session.headers["Authorization"] = "mystuff"
        iterator = iterate_jira_issues(self.fake_server.url, session=session)

        with self.assertRaises(StopIteration):
            next(iterator)

        self.assertEqual(
            self.fake_server.requests[0].headers["Authorization"], "mystuff")

    def test_no_token_or_session(self):
        """Neither a token nor a session is given"""
        with self.assertRaises(ValueError):
            next(iterate_jira_issues(self.fake_server.url, "hello"))

    def test_both_token_and_session(self):
        """Both a token and a session are given"""
        with self.assertRaises(ValueError):
            next(iterate_jira_issues(self.fake_server.url, "hello",
                                     token="mytoken",
                                     session=requests.Session()))

    def test_bad_response(self):
        """Response is not 200"""
        self.fake_server.responses.append((500, ""))
        with self.assertRaises(requests.HTTPError):
            next(iterate_jira_issues(self.fake_server.url, "hello",
                                     token="mytoken"))

    def test_repeating_header(self):
        """Header is repeated in CSV"""
        self.fake_server.responses.append((200, 'Key,Summary,Summary\n' +
                                           '"hello","world","1"\n' +
                                           '"hello","world","2"\n'))
        iterator = iterate_jira_issues(self.fake_server.url, "hello",
                                       token="mytoken")
        data = next(iterator)
        self.assertEqual(data["Summary"], ["world", "1"])
        data = next(iterator)
        self.assertEqual(data["Summary"], ["world", "2"])


class MainTestCase(TestCase):
    def test_main(self):
        """Test main function"""

        with (patch("jira2json.__main__._parse_args") as parse_args,
              patch("jira2json.__main__.iterate_jira_issues"
                    ) as iterate_jira_issues,
              patch("jira2json.__main__.parse_issues") as parse_issues,
              patch("jira2json.__main__.save_jsons_to_file"
                    ) as save_jsons_to_file):
            output = io.StringIO()
            parse_args.return_value = SimpleNamespace(jql="hello",
                                                      base_url="world",
                                                      token="mytoken",
                                                      output=output)
            iterate_jira_issues.return_value = iter([{"hello": "world"}])
            parse_issues.return_value = iter([{"hello2": "world2"}])

            _main()

            iterate_jira_issues.assert_called_once_with("world", "hello",
                                                        token="mytoken")
            parse_issues.assert_called_once_with(
                iterate_jira_issues.return_value)
            save_jsons_to_file.assert_called_once_with(
                parse_issues.return_value, output)

    def test_args(self):
        """Test argument parsing"""
        with (patch.dict(os.environ, {"JIRA_TOKEN": "tokey_the_token"})):
            args = _parse_args(
                *["--base_url", "https://jira.example.com",
                  "--jql", "project = HELLO",
                  "--token", "mytoken"])
            self.assertEqual(args.base_url, "https://jira.example.com")
            self.assertEqual(args.jql, "project = HELLO")
            self.assertEqual(args.token, "mytoken")
            self.assertEqual(args.output.name,
                             DEFAULT_FILENAME)
            args.output.close()

            args = _parse_args(
                *["--base_url", "https://jira.example.com",
                    "--jql", "project = HELLO"])

            self.assertEqual(args.token, "tokey_the_token")
            args.output.close()


if __name__ == "__main__":  # pragma: no cover
    from unittest import main
    main()


scripts/test.sh
python -m coverage run --branch tests.py && coverage html

jira2json/__init__.py
"""A module for converting JIRA issues to JSON lines format."""

__author__ = "Bar Harel"
__version__ = "0.1.1"
__license__ = "MIT License"
__all__ = ["iterate_jira_issues", "save_jsons_to_file", "parse_issues",
           "default_parsers"]

import csv as _csv
import json as _json
import logging as _logging
import typing as _typing
from collections import Counter as _Counter
from collections.abc import Iterable as _Iterable
from collections.abc import Iterator as _Iterator
from itertools import islice as _islice
from typing import Any as _Any
from typing import Callable as _Callable
from urllib.parse import urljoin as _urljoin

import requests as _requests

logger = _logging.getLogger(__name__)

# Must be <= 1000
BATCH_SIZE = 800

# JIRA API URL for CSV
CSV_API_PATH = (
    '/sr/jira.issueviews:searchrequest-csv-all-fields/temp/SearchRequest.csv')


def _log_work_parser(log_work: str | list[str] | None
                     ) -> list[dict[str, _Any]] | None:
    """Parse a log work string to a dictionary.

    Example:
        >>> _log_work_parser("comment;started;author;timeSpentSeconds")
        [{'comment': 'comment', 'started': 'started', 'author': 'author',
            'timeSpentSeconds': 'timeSpentSeconds'}]

    Args:
        log_work: A string or list of strings representing the log work entry.
            The string should be in the format of
            "comment;started;author;timeSpentSeconds". If None, an empty
            string, or an empty list, an empty dictionary is returned.

    Returns:
        A list of dictionaries representing the log work entries.
    """
    # None, empty string, or empty list
    if not log_work:
        return None

    if isinstance(log_work, str):
        log_work = [log_work]

    # Jira is stupid and doesn't escape semicolons in the log work
    values = map(lambda x: x.rsplit(";", maxsplit=3), log_work)
    return [dict(zip(("comment", "started", "author", "timeSpentSeconds"),
                     value)) for value in values]


def _comment_parser(comment: str | list[str] | None
                    ) -> list[dict[str, _Any]] | None:
    """Parse a log work string to a dictionary.

    Example:
        >>> _log_work_parser("comment;started;author;timeSpentSeconds")
        [{'comment': 'comment', 'started': 'started', 'author': 'author',
            'timeSpentSeconds': 'timeSpentSeconds'}]

    Args:
        comment: A string or list of strings representing the log work entry.
            The string should be in the format of
            "comment;started;author;timeSpentSeconds". If None, an empty
            string, or an empty list, an empty dictionary is returned.

    Returns:
        A list of dictionaries representing the log work entries.
    """
    # None, empty string, or empty list
    if not comment:
        return None

    if isinstance(comment, str):
        comment = [comment]

    # Jira is stupid and doesn't escape semicolons in the log work
    values = map(lambda x: x.split(";", maxsplit=2), comment)
    return [dict(zip(("datetime", "author", "comment"),
                     value)) for value in values]


default_parsers = {
    "Log Work": _log_work_parser,
    "Comment": _comment_parser,
}
"""A dictionary of default parsers for JIRA fields."""


def iterate_jira_issues(base_url: str, jql: str = "", *,
                        token: str | None = None,
                        session: _requests.Session | None = None
                        ) -> _Iterator[dict]:
    """Iterate over JIRA issues from the JIRA API.

    Args:
        base_url: The base URL of the JIRA server.
        jql: The JQL query to use for the search.
        token: The API token to use for authentication. If None, a session
            must be provided.
        session: A requests session to use for the API calls. If None, a
            session will be created, and the token will be used for
            authentication.

    Yields:
        A dictionary representing a single JIRA issue.
    """
    if not (bool(token) ^ bool(session)):
        raise ValueError("Either token or session must be provided.")

    if session is None:
        session = _requests.Session()
        session.headers["Authorization"] = f"Bearer {token}"

    url = _urljoin(base_url, CSV_API_PATH)
    start_at = 0
    headers = None
    single_headers = None
    response = None
    while True:

        with session.get(
                url,
                params={"jqlQuery": jql,
                        "tempMax": str(BATCH_SIZE),
                        "pager/start": str(start_at), },
                stream=True) as response:

            if not response.ok:
                logger.error(
                    "Failed to get issues from JIRA. "
                    "Status code: %s."
                    + (" Perhaps the JQL query is invalid?" if
                       response.status_code == 400 else "") +
                    " Response: %s.",
                    response.status_code, response.text)
                response.raise_for_status()

            reader = _csv.reader(response.iter_lines(
                decode_unicode=True, delimiter="\n"))
            header = next(reader, None)
            if not header:
                break
            header_count = _Counter(header)
            new_headers = set(header_count)
            new_single_headers = set(
                (key for key, value in header_count.items() if value == 1))

            if headers is not None and (new_headers != headers or
                                        new_single_headers != single_headers):
                assert single_headers is not None
                removed = headers-new_headers
                new = new_headers-headers
                changed = new | removed
                single_changed = single_headers ^ new_single_headers
                single_changed -= changed
                logger.warning("Jira fields changed between batches. Diff - "
                               "Removed fields: %s ;  New fields: %s ; "
                               "Changed from single to list or opposite: %s",
                               tuple(headers-new_headers),
                               tuple(new_headers-headers),
                               tuple(single_changed))

            single_headers = new_single_headers
            headers = new_headers

            count = 0

            for count, row in enumerate(
                    map(iter, filter(None, reader)), start=1):  # type: ignore
                output: dict = {}
                for key, value in header_count.items():
                    if value > 1:
                        output[key] = list(
                            filter(None, _islice(row, value)))  # type: ignore
                    else:
                        output[key] = next(row) or None
                yield output

            if count < BATCH_SIZE:
                break

            assert count == BATCH_SIZE, (
                "Server returned more than BATCH_SIZE issues.")

            start_at += count


def save_jsons_to_file(jsons: _Iterable[dict],
                       fileobj: _typing.TextIO) -> None:
    """Save a list of JSONs to a file.

    Args:
        jsons: An iterator of JSONable dictionaries to save.
        fileobj: A file-like object to save the JSONs to.
    """
    for json_ in jsons:
        _json.dump(json_, fileobj)
        fileobj.write("\n")


def parse_issues(
    issues: _Iterable[dict[str, _Any]], *,
    parsers: dict[str, _Callable[[_Any], _Any]] = default_parsers
) -> _Iterator[dict[str, _Any]]:
    """Applying parsers to the issues.

    Jira's data is badly formatted, and badly escaped. Apply some parsers
    to make it more usable.

    Args:
        issues: An iterator of JSONable dictionaries to parse.
        parsers: A dictionary of parsers to apply to the issues. The keys
            are the names of the fields to parse, and the values are the
            functions to apply to the values of the fields.

    Yields:
        The parsed issues.
    """
    for issue in issues:
        for key, parser in parsers.items():
            issue[key] = parser(issue.get(key))
        yield issue


jira2json/__main__.py
import argparse
import logging
import os
from typing import Any
from jira2json import iterate_jira_issues, parse_issues, save_jsons_to_file

DEFAULT_FILENAME = "jira_issues.jsonl"


def _main():
    """Entry point of the program."""

    try:
        import dotenv
        dotenv.load_dotenv()
    except ImportError:  # pragma: no cover
        pass

    logging.basicConfig(level=logging.WARNING)

    args = _parse_args()

    jsons = iterate_jira_issues(
        args.base_url, args.jql, token=args.token)
    jsons = parse_issues(jsons)

    with args.output as fileobj:
        save_jsons_to_file(jsons, fileobj)


def _parse_args(*args: str) -> argparse.Namespace:
    """Parse the command line arguments.

    Args:
        args: The command line arguments to parse. If not provided, the
            arguments are read from sys.argv.

    Returns:
        The parsed command line arguments.
    """
    parser = argparse.ArgumentParser(
        description='Convert JIRA issues to JSON format.')

    parser.add_argument(
        "--jql", type=str,
        help="The JQL query to use for the search. By default, "
        "all issues are returned.")

    def _default_environ(key: str) -> dict[str, Any]:
        """Return a dict with the default value of an environment variable.

        Args:
            key: The name of the environment variable.

        Returns:
            A dict with the default value of the environment variable, or
            {'required': True} if the environment variable is not set.
        """
        return ({'default': value} if (value := os.getenv(key)) is not None
                else {"required": True})

    parser.add_argument('-u', '--base_url', type=str,
                        **_default_environ("JIRA_BASE_URL"),
                        help='The base URL of the JIRA server. '
                        'Can also be set using the JIRA_BASE_URL environment '
                        'variable.')

    parser.add_argument("-t", "--token", type=str,
                        **_default_environ("JIRA_TOKEN"),
                        help="The API token to use for authentication. "
                        "Can also be set using the JIRA_TOKEN environment "
                        "variable.")

    parser.add_argument("-o", "--output", type=argparse.FileType("w"),
                        default=DEFAULT_FILENAME,
                        help="The output file to save the JSONs to. "
                        "By default, the JSONs are saved to "
                        "'jira_issues.jsonl.' "
                        "If the file already exists, it will be overwritten.")

    result = parser.parse_args(args or None)
    return result


if __name__ == "__main__":  # pragma: no cover
    """Entry point of the program."""
    _main()


jira_to_json.py
"""A module for converting JIRA issues to JSON lines format."""

__author__ = "Bar Harel"
__version__ = "0.1.0"
__license__ = "MIT License"
# __all__ = []

import logging
from typing import Any
from collections.abc import Iterator, Iterable, MutableMapping
import typing
import requests
from urllib.parse import urljoin, urlencode
import json
from itertools import islice
import csv
import argparse
import os
import dotenv
from collections import Counter

dotenv.load_dotenv()
logger = logging.Logger(__name__)

BATCH_SIZE = 500

# JIRA API URL for CSV
CSV_API = '/sr/jira.issueviews:searchrequest-csv-all-fields/temp/SearchRequest.csv'


def _log_work_converter(log_work: str | list[str] | None) -> list[dict[str, Any]] | None:
    """Convert a log work string to a dictionary.

    Example:
        >>> _log_work_converter("comment;started;author;timeSpentSeconds")
        [{'comment': 'comment', 'started': 'started', 'author': 'author',
            'timeSpentSeconds': 'timeSpentSeconds'}]

    Args:
        log_work: A string or list of strings representing the log work entry.
            The string should be in the format of
            "comment;started;author;timeSpentSeconds". If None, an empty
            string, or an empty list, an empty dictionary is returned.

    Returns:
        A list of dictionaries representing the log work entries.
    """
    # None, empty string, or empty list
    if not log_work:
        return None

    if isinstance(log_work, str):
        log_work = [log_work]
    
    # Jira is stupid and doesn't escape semicolons in the log work
    values = map(lambda x: x.rsplit(";", maxsplit=3), log_work)
    return [dict(zip(("comment", "started", "author", "timeSpentSeconds"),
                     value)) for value in values]

def _comment_converter(comment: str | list[str] | None) -> list[dict[str, Any]] | None:
    """Convert a log work string to a dictionary.

    Example:
        >>> _log_work_converter("comment;started;author;timeSpentSeconds")
        [{'comment': 'comment', 'started': 'started', 'author': 'author',
            'timeSpentSeconds': 'timeSpentSeconds'}]

    Args:
        comment: A string or list of strings representing the log work entry.
            The string should be in the format of
            "comment;started;author;timeSpentSeconds". If None, an empty
            string, or an empty list, an empty dictionary is returned.

    Returns:
        A list of dictionaries representing the log work entries.
    """
    # None, empty string, or empty list
    if not comment:
        return None

    if isinstance(comment, str):
        comment = [comment]
    
    # Jira is stupid and doesn't escape semicolons in the log work
    values = map(lambda x: x.split(";", maxsplit=2), comment)
    return [dict(zip(("datetime", "author", "comment"),
                     value)) for value in values]


converters = {
    "Log Work": _log_work_converter,
}


def iterate_jira_issues(base_url: str, token: str, jql: str) -> Iterator[dict]:
    """Iterate over JIRA issues from the JIRA API.

    Args:
        base_url: The base URL of the JIRA server.
        token: The API token to use for authentication.
        jql: The JQL query to use for the search.

    Yields:
        A dictionary representing a single JIRA issue.
    """
    session = requests.Session()
    session.headers["Authorization"] = f"Bearer {token}"

    url = urljoin(base_url, CSV_API)
    start_at = 0
    headers = None
    single_headers = None
    response = None
    while True:

        with session.get(
                url,
                params={"jqlQuery": jql,
                        "tempMax": BATCH_SIZE,
                        "pager/start": start_at, },
                stream=True) as response:

            if not response.ok:
                logger.error(
                    "Failed to get issues from JIRA. "
                    "Status code: %s."
                    + (" Perhaps the JQL query is invalid?" if
                       response.status_code == 400 else "") +
                    " Response: %s.",
                    response.status_code, response.text)
                response.raise_for_status()

            reader = csv.reader(response.iter_lines(
                decode_unicode=True, delimiter="\n"))
            header = next(reader, None)
            if not header:
                break
            header_count = Counter(header)
            new_headers = set(header_count)
            new_single_headers = set(
                (key for key, value in header_count.items() if value == 1))

            if headers is not None and (new_headers != headers or
                                        new_single_headers != single_headers):
                assert single_headers is not None
                removed = headers-new_headers
                new = new_headers-headers
                changed = new | removed
                single_changed = single_headers ^ new_single_headers
                single_changed -= changed
                logger.warning("Jira fields changed between batches. Diff - "
                               "Removed fields: %s ;  New fields: %s ; "
                               "Changed from single to list or opposite: %s",
                               tuple(headers-new_headers),
                               tuple(new_headers-headers),
                               tuple(single_changed))

            single_headers = new_single_headers
            headers = new_headers

            for row in map(iter, filter(None, reader)):  # type: ignore
                output: dict = {}
                for key, value in header_count.items():
                    if value > 1:
                        output[key] = list(
                            filter(None, islice(row, value)))  # type: ignore
                    else:
                        output[key] = next(row) or None
                yield output

            start_at += BATCH_SIZE


def save_jsons_to_file(jsons: Iterable[dict],
                       fileobj: typing.TextIO) -> None:
    """Save a list of JSONs to a file.

    Args:
        jsons: An iterator of JSONable dictionaries to save.
        fileobj: A file-like object to save the JSONs to.
    """
    for json_ in jsons:
        json.dump(json_, fileobj)
        fileobj.write("\n")

def convert_jsons(jsons: Iterable[dict[str, Any]]) -> Iterator[dict[str, Any]]:
    """Apply the converters to the JSONs.
    
    Args:
        jsons: An iterator of JSONable dictionaries to convert.

    Yields:
        The converted JSONs.
    """
    for json_ in jsons:
        for key, converter in converters.items():
            json_[key] = converter(json_.get(key))
        yield json_


def main():
    """Entry point of the program."""
    token = os.getenv("JIRA_TOKEN")
    if not token:
        raise ValueError("JIRA_TOKEN environment variable is not set")

    base_url = os.getenv("JIRA_BASE_URL")
    if not base_url:
        raise ValueError("JIRA_BASE_URL environment variable is not set")

    jsons = iterate_jira_issues(base_url, token, "ORDER BY WORKLOGDATE ASC")
    with open("jira_issues.jsonl", "w") as fileobj:
        save_jsons_to_file(jsons, fileobj)


if __name__ == "__main__":
    """Entry point of the program."""
    parser = argparse.ArgumentParser(
        description='Convert JIRA issues to JSON format.')

    parser.add_argument("--jql", type=str,
                        help="The JQL query to use for the search. By default, "
                        "all issues are returned.")

    def _default_environ(key: str) -> dict[str, Any]:
        """Return a dict with the default value of an environment variable.

        Args:
            key: The name of the environment variable.

        Returns:
            A dict with the default value of the environment variable, or
            {'required': True} if the environment variable is not set.
        """
        return {'default': value} if (value := os.getenv(key)) is not None else {
            "required": True}

    parser.add_argument('-u', '--base_url', type=str, **_default_environ("JIRA_BASE_URL"),
                        help='The base URL of the JIRA server. '
                        'Can also be set using the JIRA_BASE_URL environment '
                        'variable.')

    parser.add_argument("-t", "--token", type=str, **_default_environ("JIRA_TOKEN"),
                        help="The API token to use for authentication. "
                        "Can also be set using the JIRA_TOKEN environment "
                        "variable.")

    parser.add_argument("-o", "--output", type=str, default="jira_issues.jsonl",
                        help="The output file to save the JSONs to. "
                        "By default, the JSONs are saved to "
                        "'jira_issues.jsonl.' "
                        "If the file already exists, it will be overwritten.")

    args = parser.parse_args()

    jsons = iterate_jira_issues(
        args.base_url, args.token, args.jql)
    jsons = convert_jsons(jsons)

    with open(args.output, "w") as fileobj:
        save_jsons_to_file(jsons, fileobj)
